scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala

```diff
@@ -0,0 +1,268 @@
+          // and mapValues/flatMapValues on the result
+          val cp = cacheGet(hcg, mode) { implicit fd =>
+            val kvPipe = hcg.right.hashPipe(hcg.left)(hcg.joiner)(fd, mode)
+            CascadingPipe(kvPipe, kvfields, fd, mode, tuple2Converter[K, R])
+          }
+          loop(cp, rest)
+        }
+        go(hcg)
+
+      case cgp@CoGroupedPipe(_) =>
+        def go[K, V](cgp: CoGroupedPipe[K, V]): Pipe = {
+          // TODO we can push up filterKeys on both the left and right
+          // and mapValues/flatMapValues on the result
+          val cp = cacheGet(cgp, mode) { implicit fd =>
+            val kvPipe = cgp.cogrouped.toKVPipe(fd, mode)
+            CascadingPipe(kvPipe, kvfields, fd, mode, tuple2Converter[K, V])
+          }
+          loop(cp, rest)
+        }
+        go(cgp)
+
+      case r@ReduceStepPipe(_, _, _) =>
+        def go[K, V1, V2](rsp: ReduceStepPipe[K, V1, V2]): Pipe = {
+          // TODO we can push up filterKeys on both the left and right
+          // and mapValues/flatMapValues on the result
+
+          def pipe(flowDef: FlowDef) = Grouped.maybeBox[K, V1](rsp.reduce.keyOrdering, flowDef) { (tupleSetter, fields) =>
+            val (sortOpt, ts) = rsp.valueOrd.map {
+              case ordser: OrderedSerialization[V1] =>
+                // We get in here when we do a secondary sort
+                // and that sort is an ordered serialization
+                // We now need a boxed serializer for this type
+                // Then we set the comparator on the field, and finally we box the value with our tupleSetter
+                val (boxfn, boxordSer) = Grouped.getBoxFnAndOrder[V1](ordser, flowDef)
+                val valueF = new Fields("value")
+                valueF.setComparator("value", new CascadingBinaryComparator(boxordSer))
+                val ts2 = tupleSetter.asInstanceOf[TupleSetter[(K, Boxed[V1])]].contraMap { kv1: (K, V1) => (kv1._1, boxfn(kv1._2)) }
+                (Some(valueF), ts2)
+              case vs =>
+                (Some(Grouped.valueSorting(vs)), tupleSetter)
+            }.getOrElse((None, tupleSetter))
+
+            val p = rsp.reduce.mapped.toPipe(Grouped.kvFields)(flowDef, mode, TupleSetter.asSubSetter(ts))
+
+            RichPipe(p).groupBy(fields) { inGb =>
+                val withSort = sortOpt.fold(inGb)(inGb.sortBy)
+                rsp.op(withSort)
+              }
+          }
+
+          val cp = cacheGet(rsp, mode) { implicit fd =>
+            val tupConv = Grouped.tuple2Conv[K, V2](rsp.reduce.keyOrdering)
+            CascadingPipe(pipe(fd), kvfields, fd, mode, tupConv)
+          }
+          loop(cp, rest)
+      }
+      go(r)
+    }
+
+    import Dsl._ // Ensure we hook into all pipes coming out of the typed API to apply the FlowState's properties on their pipes
+    val pipe =
+      loop(p, FlatMappedFn.identity[U])
+        .applyFlowConfigProperties(flowDef)
+    RichPipe.setPipeDescriptionFrom(pipe, LineNumber.tryNonScaldingCaller)
```

Comment:
I think we should. I would like to add another AST node to annotate with line numbers rather than add it to each node. I wanted to keep this diff as small as possible though so people can read it.